### PR TITLE
[GLUTEN-2456][CH] Support use push down filter to skip rowgroups in parquet input format reader

### PR DESCRIPTION
**Pull request by KevinyhZou**
## What changes were proposed in this pull request?

(Please fill in changes proposed in this fix)

(Fixes: \#2456)

## How was this patch tested?

test by UT

### CODE REVIEW

File: cpp-ch/local-engine/Storages/SubstraitSource/ParquetFormatFile.h [hunk 1]
```diff
@@ -7,6 +7,17 @@
 #    include <IO/ReadBuffer.h>
 #    include <Storages/SubstraitSource/FormatFile.h>
 
+// clang-format off
```
Comment (taiyang-li) [inline cpp-ch/local-engine/Storages/SubstraitSource/ParquetFormatFile.h:10]: No need to set clang-format off for includes. Make sure clang-format formats all codes.

File: cpp-ch/local-engine/Storages/SubstraitSource/ParquetFormatFile.h [hunk 2]
```diff
@@ -35,9 +46,118 @@ class ParquetFormatFile : public FormatFile
 private:
     std::mutex mutex;
     std::optional<size_t> total_rows;
-
     std::vector<RowGroupInfomation> collectRequiredRowGroups(int & total_row_groups);
     std::vector<RowGroupInfomation> collectRequiredRowGroups(DB::ReadBuffer * read_buffer, int & total_row_groups);
+
+    bool checkRowGroupIfRequired(std::unique_ptr<parquet::RowGroupMetaData> meta)
```
Comment (taiyang-li) [inline cpp-ch/local-engine/Storages/SubstraitSource/ParquetFormatFile.h:52]: 函数定义放在cpp文件里

File: cpp-ch/local-engine/Storages/SubstraitSource/ParquetFormatFile.h [hunk 3]
```diff
@@ -35,9 +46,118 @@ class ParquetFormatFile : public FormatFile
...
+        return row_group_required;
+    }
+
+    void setColumnMaxMinAndType(std::shared_ptr<parquet::Statistics> statistics, 
+                    parquet::Type::type parquet_data_type, 
+                    Int32 column_type_length,
+                    std::vector<DB::Range> & max_mins,
+                    DB::DataTypes & data_types)
+    {
+        auto setColumnRangeAndType = [&](DB::Range range, DB::DataTypePtr type, std::vector<DB::Range> & ranges, DB::DataTypes & types) -> void
```
Comment (taiyang-li) [inline cpp-ch/local-engine/Storages/SubstraitSource/ParquetFormatFile.h:99]: follow code style of CH, setColumnRangeAndType -> set_column_range_and_type

File: cpp-ch/local-engine/Storages/SubstraitSource/ParquetFormatFile.h [hunk 4]
```diff
@@ -35,9 +46,118 @@ class ParquetFormatFile : public FormatFile
...
+        {
+            ranges.push_back(range);
+            types.push_back(type);
+        };
```
Comment (taiyang-li) [inline cpp-ch/local-engine/Storages/SubstraitSource/ParquetFormatFile.h:103]: proper blank line

File: cpp-ch/local-engine/Storages/SubstraitSource/FormatFile.h [hunk 5]
```diff
@@ -58,13 +59,20 @@ class FormatFile
 
     virtual size_t getStartOffset() const { return file_info.start(); }
     virtual size_t getLength() const { return file_info.length(); }
+    
+    void setFilters(std::vector<DB::KeyCondition> & filters_)
```
Comment (taiyang-li) [inline cpp-ch/local-engine/Storages/SubstraitSource/FormatFile.h:63]: 输入参数加const

File: cpp-ch/local-engine/Storages/SubstraitSource/FormatFile.h [hunk 6]
```diff
@@ -58,13 +59,20 @@ class FormatFile
...
+    { 
+        for (size_t i = 0; i < filters_.size(); ++i)
+            filters.push_back(filters_[i]);
```
Comment (taiyang-li) [inline cpp-ch/local-engine/Storages/SubstraitSource/FormatFile.h:66]: std::vector赋值直接用 filters = filters_;

File: cpp-ch/local-engine/Storages/SubstraitSource/ParquetFormatFile.cpp [hunk 7]
```diff
@@ -16,6 +16,24 @@
...
+#include <Formats/FormatSettings.h>
+#include <IO/SeekableReadBuffer.h>
+#include <Storages/ArrowParquetBlockInputFormat.h>
+#include <Processors/Formats/Impl/ArrowBufferedStreams.h>
+#include <Processors/Formats/Impl/ParquetBlockInputFormat.h>
+#include <Processors/Formats/Impl/ArrowColumnToCHColumn.h>
+
+#include <DataTypes/DataTypesNumber.h>
+
+// clang-format on
```
Comment (taiyang-li) [inline cpp-ch/local-engine/Storages/SubstraitSource/ParquetFormatFile.cpp:36]: 除非特殊需要，最好不要随意开闭clang-format

**Review by taiyang-li — CHANGES_REQUESTED** [review]

File: cpp-ch/local-engine/Parser/SerializedPlanParser.cpp [hunk 8]
```diff
@@ -246,8 +247,15 @@ QueryPlanStepPtr SerializedPlanParser::parseReadRealWithLocalFile(const substrai
     auto header = TypeParser::buildBlockFromNamedStruct(rel.base_schema());
     auto source = std::make_shared<SubstraitFileSource>(context, header, rel.local_files());
     auto source_pipe = Pipe(source);
-    auto source_step = std::make_unique<ReadFromStorageStep>(std::move(source_pipe), "substrait local files", nullptr);
+    // auto source_step = std::make_unique<ReadFromStorageStep>(std::move(source_pipe), "substrait local files", nullptr);
```
Comment (taiyang-li) [inline cpp-ch/local-engine/Parser/SerializedPlanParser.cpp:250]: remove useless codes

File: cpp-ch/local-engine/Storages/SubstraitSource/ParquetFormatFile.cpp [hunk 9]
```diff
@@ -16,6 +16,24 @@
 #    include <parquet/arrow/reader.h>
 #    include <Common/Config.h>
 
+#include <memory>
```
Comment (taiyang-li) [inline cpp-ch/local-engine/Storages/SubstraitSource/ParquetFormatFile.cpp:19]: 重复include, 建议打开clang-format, 格式化下所有#include

File: cpp-ch/local-engine/Storages/SubstraitSource/ParquetFormatFile.cpp [hunk 10]
```diff
@@ -48,7 +66,6 @@ FormatFile::InputFormatPtr ParquetFormatFile::createInputFormat(const DB::Block
     }
     else
         required_row_groups = collectRequiredRowGroups(total_row_groups);
-
```
Comment (taiyang-li) [inline cpp-ch/local-engine/Storages/SubstraitSource/ParquetFormatFile.cpp:51]: 代码段之间适当保留的空行，这里没必要去掉。

**Review by taiyang-li — CHANGES_REQUESTED** [review]

File: cpp-ch/local-engine/Storages/SubstraitSource/ParquetFormatFile.cpp [hunk 11]
```diff
@@ -157,5 +175,117 @@ std::vector<RowGroupInfomation> ParquetFormatFile::collectRequiredRowGroups(DB::
...
+    DB::DataTypes column_types;
+    const parquet::SchemaDescriptor* schema_desc = meta->schema();
+    int num_columns = meta->num_columns();
+    for (int i = 0; i < num_columns; ++i)
+    {
+        const parquet::ColumnDescriptor * desc = schema_desc->Column(i);
+        Int32 column_type_length = desc->type_length();
+        auto column_chunk_meta = meta->ColumnChunk(i);
+        if (column_chunk_meta->is_stats_set())
+            setColumnMaxMinAndType(column_chunk_meta->statistics(), column_chunk_meta->type(), column_type_length, column_max_mins, column_types);
```
Comment (taiyang-li) [inline cpp-ch/local-engine/Storages/SubstraitSource/ParquetFormatFile.cpp:191]: 从设计上来讲，这里最好不要将std::vector::push_back的逻辑和setColumnMaxMinAndType耦合，setColumnMaxMinAndType只返回DB::Range，再在setColumnMaxMinAndType外面将DB::Range追加到std::vector<DB::Range>中。方便后续扩展。

**Review by taiyang-li — CHANGES_REQUESTED** [review]

File: cpp-ch/local-engine/Storages/SubstraitSource/FormatFile.h [hunk 12]
```diff
@@ -74,13 +75,17 @@ class FormatFile
 
     virtual size_t getStartOffset() const { return file_info.start(); }
     virtual size_t getLength() const { return file_info.length(); }
-
+    
+    void setFilters(std::vector<SourceFilter> filters_) { filters = std::move(filters_); }
+    virtual String getFileFormat() const { return "unknown"; }
```
Comment (taiyang-li) [inline cpp-ch/local-engine/Storages/SubstraitSource/FormatFile.h:80]: 这是必须由派生类实现的接口，定义为纯虚函数较妥

File: cpp-ch/local-engine/Storages/SubstraitSource/ParquetFormatFile.cpp [hunk 13]
```diff
@@ -32,11 +32,30 @@
 #    include <parquet/arrow/reader.h>
 #    include <Common/Config.h>
 
+#    include <memory>
+#    include <string>
+#    include <utility>
```
Comment (taiyang-li) [inline cpp-ch/local-engine/Storages/SubstraitSource/ParquetFormatFile.cpp:37]: image

删掉中间空行，使用clang-format重新格式化，它会自动去重。

**Review by taiyang-li — CHANGES_REQUESTED** [review]

File: cpp-ch/local-engine/Storages/SubstraitSource/ParquetFormatFile.cpp [hunk 14]
```diff
@@ -173,5 +194,110 @@ std::vector<RowGroupInfomation> ParquetFormatFile::collectRequiredRowGroups(DB::
...
+        auto iter = filters[i].keys.begin();
+
+        for (size_t j = 0; j < filters[i].keys.size(); ++j)
+        {
+            DB::String filter_col_key = iter->name;
+            DB::DataTypePtr filter_col_type = iter->type;
+            int column_index = schema_desc->ColumnIndex(filter_col_key);
+            DB::Range range = DB::Range::createWholeUniverse();
+            if (column_index < 0)
+                ranges.push_back(range);
```
Comment (taiyang-li) [inline cpp-ch/local-engine/Storages/SubstraitSource/ParquetFormatFile.cpp:216]: ranges.emplace_back(std::move(range));

File: cpp-ch/local-engine/Storages/SubstraitSource/SubstraitFileSourceStep.cpp [hunk 15]
```diff
@@ -0,0 +1,158 @@
...
+            }
+            else
+            {
+                DB::NameAndTypePair name_and_type(column.name, column.type);
+                filter_column_keys.push_back(name_and_type);
+            }
+        }
+    }
+
+    if (filter_column_keys.size() != 0)
```
Comment (taiyang-li) [inline cpp-ch/local-engine/Storages/SubstraitSource/SubstraitFileSourceStep.cpp:140]: if (!filter_column_keys.empty())

File: cpp-ch/local-engine/Storages/SubstraitSource/SubstraitFileSource.cpp [hunk 16]
```diff
@@ -121,6 +122,25 @@ SubstraitFileSource::SubstraitFileSource(
     }
 }
 
+void SubstraitFileSource::applyFilters(std::vector<SourceFilter> filters_) const
+{
+    for (size_t i=0; i < files.size(); ++i)
```
Comment (taiyang-li) [inline cpp-ch/local-engine/Storages/SubstraitSource/SubstraitFileSource.cpp:127]: 格式化

**Review by taiyang-li — CHANGES_REQUESTED** [review]

**Comment by KevinyhZou** [comment]
Fixed done, pelase review the code again @taiyang-li

**Review by taiyang-li — APPROVED** [review]
LGTM

**Review by baibaichen — APPROVED** [review]
LGTM

**Review by taiyang-li — APPROVED** [review]
LGTM

**Review by baibaichen — APPROVED** [review]
LGTM

**Pull request merged**